### PR TITLE
Persistent registry

### PR DESCRIPTION
**Pull request by adam-yorwerth**
Replacing the in-memory implementation of the node registry with a Postgres version.

This will mean that between deployments and restarts the state of the registry is maintained.

### CODE REVIEW

File: registry-http-server/src/main/java/com/tesco/aqueduct/registry/NodeRegistryController.java [hunk 1]
```diff
@@ -3,13 +3,13 @@
 import com.tesco.aqueduct.pipe.api.MessageReader;
 import com.tesco.aqueduct.pipe.metrics.Measure;
 import io.micronaut.http.annotation.*;
+import io.micronaut.retry.annotation.Retryable;
```
Comment (cwebbtw) [inline registry-http-server/src/main/java/com/tesco/aqueduct/registry/NodeRegistryController.java]: Unused import here, optimise imports to remove :-)

Comment (adam-yorwerth) [reply registry-http-server/src/main/java/com/tesco/aqueduct/registry/NodeRegistryController.java]: Addressed

File: registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+        ZonedDateTime threshold = ZonedDateTime.now().minus(offlineDelta);
+
+        if (node.getLastSeen().compareTo(threshold) < 0) {
+            return node.toBuilder().status("offline").build();
+        }
+        return node;
+    }
+
+    private List<NodeGroup> getNodesFilteredByGroup(Connection connection, List<String> groups) throws SQLException {
+        List<NodeGroup> list = new ArrayList<>();
```
Comment (cwebbtw) [inline registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java]: possible use of stream->map here

File: registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+            NodeGroup nodeGroup = getNodeGroup(connection, group);
+            list.add(nodeGroup);
+        }
+        return list;
+    }
+
+    // Methods that interact with the database - possibly need to split these into another class
+
+    private NodeGroup getNodeGroup(Connection connection, String group) throws SQLException {
+        PreparedStatement statement = connection.prepareStatement(getNodeGroupQuery());
```
Comment (cwebbtw) [inline registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java]: Make sure you close prepared statements and result sets or you will be leaking.

File: registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+        statement.setString(1, group);
+
+        List<Node> nodes = new ArrayList<>();
+        int version = 0;
+
+        try (ResultSet rs = statement.executeQuery()) {
+            while (rs.next()) {
+                String entry = rs.getString("entry");
+                version= rs.getInt("version");
```
Comment (cwebbtw) [inline registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java]: space here

File: registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+                throw new RuntimeException(exception);
+            } catch (VersionChangedException exception) {
+                try {
+                    Thread.sleep(10);
+                    completed = false;
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        } while (!completed);
```
Comment (KrzysztofKowalczyk) [inline registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java]: This is an infinite loop so it should look like one

File: build.gradle [hunk 6]
```diff
@@ -29,6 +29,9 @@ apply plugin: "pl.allegro.tech.build.axion-release"
 apply plugin: 'org.jetbrains.gradle.plugin.idea-ext'
 
 dependencyCheck {
+
+    failOnError = false
```
Comment (adam-yorwerth) [inline build.gradle]: Added as the OWASP scan has started to become flaky and non-deterministically fail.

File: pipe-http-server-cloud/src/integration/NodeRegistryControllerIntegrationSpec.groovy [hunk 7]
```diff
@@ -1,22 +1,28 @@
...
+import javax.sql.DataSource
+import java.sql.DriverManager
 import java.time.Duration
 
 import static io.restassured.RestAssured.given
 import static io.restassured.RestAssured.when
 import static org.hamcrest.Matchers.*
 
-class NodeRegistryControllerSpec extends Specification {
+class NodeRegistryControllerIntegrationSpec extends Specification {
```
Comment (adam-yorwerth) [inline pipe-http-server-cloud/src/integration/NodeRegistryControllerIntegrationSpec.groovy]: Renamed as it is now at Integration level

File: registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java (hunk 2 above)
Comment (adam-yorwerth) [reply registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java]: Tried this, ends up with an additional try-catch that makes the code less readable.

File: registry-core/src/integration/groovy/com/tesco/aqueduct/registry/PostgreSQLNodeRegistryIntegrationSpec.groovy [hunk 8]
```diff
@@ -227,21 +268,98 @@ class NodeRegistrySpec extends Specification {
...
             }
         }
 
+        startLock.complete(true)
+
         then: "summary of the registry is as expected"
-        for(int i =0 ; i<50; i++) {
-            registry.getSummary(0, "initialising", ["x" + i % 50]).followers.size() == 1000
+        conditions.eventually {
+            assert registry.getSummary(0, "initialising", ["x"]).followers.size() == tills
```
Comment (KrzysztofKowalczyk) [inline registry-core/src/integration/groovy/com/tesco/aqueduct/registry/PostgreSQLNodeRegistryIntegrationSpec.groovy]: Tests asserting something are better 👍

File: registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+            statement.setString(2, jsonNodes);
+
+            return statement.executeUpdate() != 0;
+        }
+    }
+
+    private List<Node> getAllNodes(Connection connection) throws SQLException {
+        List<Node> nodes;
+        try (PreparedStatement statement = connection.prepareStatement(getAllNodesQuery())) {
+
```
Comment (cwebbtw) [inline registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java]: *nitpick* new line here looks a bit funny

File: registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+                "AND " +
+                    "registry.version = ? " +
+                ";";
+    }
+
+    private String getInsertGroupQuery() {
+        return "INSERT INTO registry (group_id, entry, version)" +
+                "VALUES (" +
+                "?, " +
+                "?::JSON , " +
```
Comment (cwebbtw) [inline registry-core/src/main/java/com/tesco/aqueduct/registry/PostgreSQLNodeRegistry.java]: I really dislike that comma being spaced between the JSON.

**Pull request merged**